### PR TITLE
Persistence

### PR DESCRIPTION
**Pull request by vishakha041**
These are the updates based on review comments. The last two commits will be merged with the second commit. So a total of 3 commits when we merge this branch.

My non-scientific evaluation: For our run_all.sh timing when running the unit tests (time in seconds).

Without a map
MsyncOnCommit - 114.555903399
NoMsync - 1.764198926
AlwaysMsync - 209.394729821

With a map
MsyncOnCommit - 106.721010523
NoMsync - 1.843715119
AlwaysMsync - 206.935488722

just clflush and sfence: 2.263883903

### CODE REVIEW

**Comment by luisremis** [comment]
It looks good to me, I asked Philip to explain a couple of things to me, and he is going over the updates.

You can only squash ALL the commits without new approval. If we approve and then you squash some of the commits, we need to approve again for the merge.

File: src/arch.h [hunk 1]
```diff
@@ -30,6 +30,7 @@
 #pragma once
 
 #include <stdint.h>
+#include "os.h"
```
Comment (philiplantz) [inline src/arch.h]: This is not needed since the calls to os have been moved to TransactionManager.

File: src/arch.h [hunk 2]
```diff
@@ -73,52 +74,17 @@ static inline void memory_barrier() // Instruct compiler not to re-order
...
+#define CONCAT(x, y)      CONCAT_(x, y)
+#define STRINGIFY_(x)     #x
+#define STRINGIFY(x)      STRINGIFY_(x) 
 
 static inline void clflush(void *addr)
 {
-#ifndef NOPM
-    __asm__("clflushopt \"%0\"" : "+m"(*(char *)addr) : : "memory");
-#endif
+    __asm__(STRINGIFY(PMFLUSH) " %0" : : "m"(*(char (*)[64])((uintptr_t)addr & -64)));
```
Comment (philiplantz) [inline src/arch.h]: `__asm__(STRINGIFY(PMFLUSH) " %0" : : "m"(*(char *)addr), "m"(*(char (*)[64])((uintptr_t)addr & -64)));`

This avoids the compiler having to generate extra code to compute the address.

File: src/graph.cc [hunk 3]
```diff
@@ -137,18 +137,22 @@ GraphImpl::GraphInit::GraphInit(const char *name, int options,
...
     : create(options & Graph::Create),
       read_only(options & Graph::ReadOnly),
-      msync_options(options & Graph::AlwaysMsync),
       info_map(name, info_name,
                GraphConfig::BASE_ADDRESS, GraphConfig::INFO_SIZE,
                create, false, read_only),
       info(reinterpret_cast<GraphInfo *>(GraphConfig::BASE_ADDRESS))
 {
+    int msync_options = options & Graph::AlwaysMsync;
+    msync_options = msync_options == 0 ? Graph::MsyncOnCommit : msync_options;
```
Comment (philiplantz) [inline src/graph.cc]:
````
switch (msync_options) {
    case 0:
    case Graph::MsyncOnCommit:
        msync_flush = true;
        always_msync = false;
        break;

    case Graph::AlwaysMsync:
        msync_flush = always_msync = true;
        break;

    case Graph::NoMsync:
        msync_flush = always_msync = false;
        break;
}
```
`

File: src/GraphImpl.h [hunk 4]
```diff
@@ -50,6 +50,8 @@ namespace PMGD {
         struct GraphInit {
             bool create;
             bool read_only;
+            bool is_msync_needed; // false only when NoMsync used for msync cases.
```
Comment (philiplantz) [inline src/GraphImpl.h]: This should be called msync_needed, to use a consistent name everywhere.

File: src/GraphImpl.h [hunk 5]
```diff
@@ -103,5 +105,11 @@ namespace PMGD {
             if (_init.read_only)
                 throw PMGDException(ReadOnly);
         }
+
+        void msync_options(bool &msync_needed, bool &sync)
+        {
```
Comment (philiplantz) [inline src/GraphImpl.h]: 'sync' should be called 'always_msync', to use a consistent name everywhere.

File: src/TransactionImpl.h [hunk 6]
```diff
@@ -49,6 +49,8 @@ namespace PMGD {
 
             GraphImpl *_db;
             int _tx_type;
+            bool _msync_needed;    // false only when NoMsync used for msync cases.
+            bool _always_sync;  // true only when AlwaysMsync used for msync cases.
```
Comment (philiplantz) [inline src/TransactionImpl.h]: This should be called always_msync, to use a consistent name everywhere.

File: src/TransactionManager.cc [hunk 7]
```diff
@@ -119,7 +120,7 @@ void TransactionManager::recover(bool read_only)
     _cur_tx_id = max_tx_id;
 }
 
-TransactionHandle TransactionManager::alloc_transaction(bool read_only)
+TransactionHandle TransactionManager::alloc_transaction(bool read_only, const bool msync_needed)
```
Comment (philiplantz) [inline src/TransactionManager.cc]: Don't use 'const' on parameters that aren't pointers or references.

File: src/TransactionManager.h [hunk 8]
```diff
@@ -73,8 +77,8 @@ namespace PMGD {
         size_t _extent_size;
         int _max_extents;
 
-        void reset_table();
-        void recover(bool read_only);
+        void reset_table(const bool msync_needed);
+        void recover(bool read_only, const bool msync_needed);
```
Comment (philiplantz) [inline src/TransactionManager.h]: Don't use 'const' on parameters that aren't pointers or references.

File: src/TransactionManager.h [hunk 9]
```diff
@@ -86,9 +90,34 @@ namespace PMGD {
                            uint64_t transaction_table_size,
                            uint64_t journal_addr,
                            uint64_t journal_size,
-                           bool create, bool read_only);
+                           bool create, bool read_only,
+                           const bool msync_needed);
```
Comment (philiplantz) [inline src/TransactionManager.h]: Don't use 'const' on parameters that aren't pointers or references.

File: src/transaction.cc [hunk 10]
```diff
@@ -167,26 +169,32 @@ static inline T align_high(T var, size_t sz)
     return (T)(((uint64_t)var + (sz-1)) & ~(sz-1));
 }
 
-void TransactionImpl::flush_range(void *ptr, size_t len)
+void TransactionImpl::flush_range(void *ptr, size_t len, bool is_msync_needed)
```
Comment (philiplantz) [inline src/transaction.cc]: This should be called msync_needed, to use a consistent name everywhere.

File: src/transaction.cc [hunk 11]
```diff
@@ -167,26 +169,32 @@ static inline T align_high(T var, size_t sz)
...
 {
     JournalEntry *jend = static_cast<JournalEntry *>(h.jend);
-    rollback(h, jend);
+    rollback(h, jend, msync_needed);
 }
 
 void TransactionImpl::rollback(const TransactionHandle &h,
-                               const JournalEntry *jend)
+                               const JournalEntry *jend,
+                               bool is_msync_needed)
```
Comment (philiplantz) [inline src/transaction.cc]: This should be called msync_needed, to use a consistent name everywhere.

File: src/transaction.cc [hunk 12]
```diff
@@ -197,7 +205,10 @@ void TransactionImpl::rollback(const TransactionHandle &h,
...
     while (je-- > jbegin) {
         memcpy(je->addr, &je->data[0], je->len);
-        clflush(je->addr);
+        TransactionManager::flush(je->addr, is_msync_needed);
     }
-    persistent_barrier();
+
+    // Unless we have NoMsync, rollback can safely commit in case
+    // of PM=MSYNC
+    TransactionManager::commit(is_msync_needed);
```
Comment (philiplantz) [inline src/transaction.cc]: When rollback is called from recover, there is no need to do msync after rolling back each transaction.

An improvement would be to remove the call to commit from both here and finalize_commit, and add a call to commit in ~TransactionImpl  and in TransactionManager::recover (after rolling back all transactions).

**Pull request merged**